<a href="https://colab.research.google.com/github/pujisubarkah/replikasi_DAiSEE-Towards-User-Engagement-Recognition-in-the-Wild/blob/main/notebook_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 2

## Video Preprocessing

Paper Replication

DAiSEE: Towards User Engagement Recognition in the Wild

In [ ]:
import os
import cv2
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from pathlib import Path

In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("olgaparfenova/daisee")

print("Path to dataset files:", path)

100%|██████████| 14.3G/14.3G [02:29<00:00, 103MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/olgaparfenova/daisee/versions/1


In [ ]:
import os

DAISEE_PATH = os.path.join(path, "DAiSEE")

DATA_PATH = os.path.join(DAISEE_PATH, "DataSet")
LABEL_PATH = os.path.join(DAISEE_PATH, "Labels")

TRAIN_PATH = os.path.join(DATA_PATH, "Train")
VAL_PATH   = os.path.join(DATA_PATH, "Validation")
TEST_PATH  = os.path.join(DATA_PATH, "Test")

In [ ]:
train_labels = pd.read_csv(
    os.path.join(LABEL_PATH, "TrainLabels.csv")
)

train_labels.head()

,ClipID,Boredom,Engagement,Confusion,Frustration
0,1100011002.avi,0,2,0,0
1,1100011003.avi,0,2,0,0
2,1100011004.avi,0,3,0,0
3,1100011005.avi,0,3,0,0
4,1100011006.avi,0,3,0,0


In [ ]:
print(f"Contents of {subject_path}:")
print(os.listdir(subject_path))

Contents of /root/.cache/kagglehub/datasets/olgaparfenova/daisee/versions/1/DAiSEE/DataSet/Train/110001:
['1100012061', '1100012046', '1100011020', '1100012059', '1100011016', '1100011080', '1100012021', '1100011049', '1100011062', '1100012052', '1100011015', '1100012010', '1100011006', '1100011059', '1100011040', '1100012026', '1100012003', '1100012045', '1100011046', '1100011026', '1100011073', '1100011003', '1100012007', '1100011013', '1100011037', '1100011005', '1100012057', '1100011021', '1100011055', '1100011018', '1100012018', '1100011067', '1100012065', '1100011079', '1100012038', '1100012022', '1100012023', '1100012030', '1100011002', '1100011034', '1100011078', '1100011010', '1100012001', '1100012041', '1100011017', '1100012028', '1100011019', '1100011068', '1100011050', '1100012032', '1100011081', '1100011072', '1100011058', '1100012050', '1100012047', '1100012060', '1100012051', '1100012062', '1100011048', '1100012008', '1100011031', '1100011082', '1100012063', '1100012011'

In [ ]:
subject = "110001"

subject_path = os.path.join(TRAIN_PATH, subject)

clip_dirs = sorted([
    d for d in os.listdir(subject_path)
    if os.path.isdir(os.path.join(subject_path, d))
])

selected_clip = clip_dirs[0]

clip_path = os.path.join(subject_path, selected_clip)

videos = sorted([
    f for f in os.listdir(clip_path)
    if f.endswith(".avi")
])

video_path = os.path.join(clip_path, videos[0])

print(video_path)

/root/.cache/kagglehub/datasets/olgaparfenova/daisee/versions/1/DAiSEE/DataSet/Train/110001/1100011002/1100011002.avi


In [ ]:
#load video
def load_video(video_path):

    cap = cv2.VideoCapture(video_path)

    frames = []

    while True:

        ret, frame = cap.read()

        if not ret:
            break

        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

        frames.append(frame)

    cap.release()

    return frames

In [ ]:
frames = load_video(video_path)

print("Jumlah frame:", len(frames))

Jumlah frame: 300


In [ ]:
def uniform_sampling(frames, num_frames=30):

    indices = np.linspace(
        0,
        len(frames)-1,
        num_frames,
        dtype=int
    )

    sampled = [frames[i] for i in indices]

    return sampled

In [ ]:
sampled_frames = uniform_sampling(frames)

print(len(sampled_frames))

30


In [ ]:
def resize_frames(frames, size=(224,224)):

    return [
        cv2.resize(frame, size)
        for frame in frames
    ]

In [ ]:
def normalize_frames(frames):

    frames = np.array(frames, dtype=np.float32)

    frames /= 255.0

    return frames

In [ ]:
def preprocess_video(video_path):

    frames = load_video(video_path)

    frames = uniform_sampling(frames)

    frames = resize_frames(frames)

    frames = normalize_frames(frames)

    return frames

In [ ]:
video = preprocess_video(video_path)

print(video.shape)

(30, 224, 224, 3)


## Ringkasan Progres Notebook

Kami telah berhasil menyelesaikan tugas-tugas berikut:

1.  **Pengaturan Lingkungan**: Mengimpor pustaka penting seperti (`os`, `cv2`, `numpy`, `pandas`, `matplotlib.pyplot`, `pathlib`).
2.  **Unduh Dataset**: Dataset DAiSEE berhasil diunduh dan diekstrak dari `kagglehub` ke `/root/.cache/kagglehub/datasets/olgaparfenova/daisee/versions/1`.
3.  **Definisi Jalur**: Jalur direktori utama telah didefinisikan (`DAISEE_PATH`, `DATA_PATH`, `LABEL_PATH`, `TRAIN_PATH`, `VAL_PATH`, `TEST_PATH`) untuk akses mudah ke komponen dataset.
4.  **Pemuatan Label**: File `TrainLabels.csv` dimuat ke dalam DataFrame pandas dan baris awalnya ditampilkan.
5.  **Pemilihan Video**: Berhasil memilih subjek spesifik (`110001`) dan menemukan klip video pertama (`1100011002`) dalam direktorinya. Ini melibatkan navigasi struktur direktori bertingkat dengan benar untuk menemukan jalur lengkap ke file video (`/root/.cache/kagglehub/datasets/olgaparfenova/daisee/versions/1/DAiSEE/DataSet/Train/110001/1100011002/1100011002.avi`), menyelesaikan `IndexError` sebelumnya.
6.  **Fungsi Pra-pemrosesan Video**: Mengembangkan serangkaian fungsi modular untuk pra-pemrosesan video:
    *   `load_video(video_path)`: Membaca semua bingkai dari video menggunakan OpenCV dan mengonversinya ke format RGB.
    *   `uniform_sampling(frames, num_frames=30)`: Mengambil sejumlah bingkai tetap (standar 30) secara seragam di seluruh video.
    *   `resize_frames(frames, size=(224,224))`: Mengubah ukuran semua bingkai ke resolusi yang ditentukan (standar 224x224 piksel).
    *   `normalize_frames(frames)`: Mengonversi bingkai ke array NumPy float32 dan menormalisasi nilai piksel ke rentang [0, 1].
7.  **Eksekusi Pra-pemrosesan Video**: Menerapkan fungsi `preprocess_video(video_path)`, yang mengelola langkah-langkah di atas, ke video yang dipilih. Data video yang telah dipra-proses memiliki bentuk `(30, 224, 224, 3)`, menunjukkan 30 bingkai, masing-masing 224x224 piksel dengan 3 saluran warna, siap untuk analisis lebih lanjut atau masukan model.

Ini melengkapi pengaturan awal dan pra-pemrosesan video dari dataset DAiSEE.